### PR TITLE
[Inductor cutlass backend] Enable StreamK and nonzero workspace sizes

### PR DESCRIPTION
**Pull request by kadeng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119010
* #119009
* #119008
* #119007
* #119006
* __->__ #119005
* #119004
* #119003

This change introduces temporary workspace buffers, which can be used to
provide workspace memory to Inductor ops. This is a required feature
in order to enable the Cutlass StreamK tile scheduler, which should
bring performance benefits and more reliable GEMM performance for
many tile shapes.

See StreamK paper: https://arxiv.org/abs/2301.03598

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -3028,6 +3028,41 @@ class InputBuffer(Buffer):
     pass
 
 
+@dataclasses.dataclass
+class WorkspaceBuffer(Buffer):
```
Comment (peterbell10) [inline torch/_inductor/ir.py:3032]: In #117992 I also add workspace allocations but take a slightly different approach. The workspace size is an implementation detail of the code generation, so I don't represent it in the inductor IR and instead put it in `KernelArgs` and have the `Kernel` method create the args.

wdyt?

Comment (kadeng) [reply torch/_inductor/ir.py:3032]: @peterbell10 - Just looked at it, looks good to me. If the PR you mention is blocked through this ROCm issue (Is that blocking?),maybe you could split the workspace allocation code off into a separate PR, so we can get it merged quickly and I can adapt this to it.

**Comment by kadeng** [comment]
Had to be squashed into https://github.com/pytorch/pytorch/pull/119007

File: torch/_inductor/ir.py (hunk 1 above)
Comment (peterbell10) [reply torch/_inductor/ir.py:3032]: ROCm issues are fixed and I'm just waiting on perf results before merging.

**Pull request closed without merging**